### ISSUE TITLE
twisted.python.deprecate does the wrong thing for methods

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3254]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3254|
|Type|defect|
|Created|2008-05-29 13:28:53Z|
|Branch|https://github.com/twisted/twisted/tree/fqn-3254-2|

If `deprecated()` is applied to a function which is a method of a class, the resulting deprecation warning will mention the name `module.function` instead of `module.class.function`.
 

<details><summary>Searchable metadata</summary>

```
trac-id__3254 3254
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_fqn_3254_2 branches/fqn-3254-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1212067733000000 1212067733000000
changetime__1222520000000000 1222520000000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [23771]) Branching to 'fqn-3254'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [24294]) Branching to 'fqn-3254-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

Let's move on this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

  * Expanding the docstring for `fullyQualifiedName` a bit would probably be good.  It says "of an object" - what kind of objects does it mean, though?  Classes, functions, methods, and modules, it seems.  Also, please document the return type.
  * Instead of `test_nestedClass` with a `todo`, how about a low-priority enhancement ticket with that test in its description?  Someone might resolve a ticket, but I can't remember the last time I saw someone fix a todo.
  * `staticmethod` and `classmethod` aren't supported.  Don't know if I much care.  Maybe we can wait until we want to deprecate a `staticmethod` or a `classmethod`.

Interesting that you chose to add a new function instead of modifying `qual`.  Any particular reason for that?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

I fixed the docstring and removed the todo. I'll create a ticket for the other problems once merged.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

  * The `return` in `_checkFullyQualfiedName` doesn't seem to do anything useful
  * `DummyClass` is still defined but no longer used

Otherwise looks good, please merge when you're happy with it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [24931]) Merge fqn-3254-2

Author: therve
Reviewer: exarkun
Fixes [#3254](https://github.com/twisted/twisted/issues/3254)

Add function fullyQualifiedName which is able to get the name of a module,
class, method or function, and use it in the deprecation mechanism instead of
qual: this makes deprecation on method return the correct name.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
